keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,723 @@
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.
+Users can still request to delete its namespace or other objects in the namespace.
+When deletion of the namespace A is requested, the deletion isn't completed until the object with lien is deleted.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Per Controller `Liens` vs Per Reason `Liens`
+
+The same controllers may need to add `Liens` to a certain resource for multiple reasons, like multiple resources are using it.
+In such a case, if `Liens` are added per controller basis, such a controller needs to be sure that no other resources are actually using the resource, when removing its `Liens` from the resource.
+On the other hand, if `Liens` are added per reason basis, such a controller can remove the `Lien` for the resource, without considering other resources.
+Also, the latter way helps users to know which particular resources are using the resource.
+
+`Liens` allows both per controller `Liens` and per reason `Liens`, and reason may exist per object.
+Only requirements are:
+  - Use key as identity of controllers or users that set the lien, and only add/remove its own list of liens
+  - `lienID` is the only mandatory field in list of liens, and it needs to be unique across the lien for a key
+
+Examples of the use of lien are as follows:
+
+Let's assume that a secret protection controller would like to protect Secret `my-secret` in `ns1` when
+both PersistentVolume `my-pv1` and Pod `my-pod1` in namespace `ns2` are using it.
+
+- Per Controller `Liens`:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: ns1
+  name: my-secret
+  liens:
+    k8s.io/secret-protection-controller:
+      - lienID: used
+```
+
+- Per Reason `Liens`:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: ns1
+  name: my-secret
+  liens:
+    k8s.io/secret-protection-controller:
+      - lienID: used by PV(s)
+      - lienID: used by Pod(s)
+```
+
+- Per Object `Liens`:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: ns1
+  name: my-secret
+  liens:
+    k8s.io/secret-protection-controller:
+      - lienID: v1/PersistentVolume//my-pv1
```

Comment:
Wearing my scalability hat, I would like to ensure that this won't be growing the object too far.

Reply:
I commented a bit more on that above.